lib/segment/src/vector_storage/multi_dense/appendable_mmap_multi_dense_vector_storage.rs

```diff
@@ -0,0 +1,271 @@
+use std::ops::Range;
+use std::path::{Path, PathBuf};
+use std::sync::atomic::AtomicBool;
+use std::sync::Arc;
+
+use atomic_refcell::AtomicRefCell;
+use bitvec::prelude::BitSlice;
+use common::types::PointOffsetType;
+
+use crate::common::operation_error::{check_process_stopped, OperationResult};
+use crate::common::Flusher;
+use crate::data_types::named_vectors::CowVector;
+use crate::data_types::primitive::PrimitiveVectorElement;
+use crate::data_types::vectors::{
+    MultiDenseVector, TypedMultiDenseVector, TypedMultiDenseVectorRef, VectorElementType, VectorRef,
+};
+use crate::types::{Distance, MultiVectorConfig, VectorStorageDatatype};
+use crate::vector_storage::chunked_mmap_vectors::ChunkedMmapVectors;
+use crate::vector_storage::dense::dynamic_mmap_flags::DynamicMmapFlags;
+use crate::vector_storage::{MultiVectorStorage, VectorStorage, VectorStorageEnum};
+
+const VECTORS_DIR_PATH: &str = "vectors";
+const OFFSETS_DIR_PATH: &str = "offsets";
+const DELETED_DIR_PATH: &str = "deleted";
+
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+struct MultivectorMmapOffset {
+    offset: PointOffsetType,
+    count: PointOffsetType,
+    reserved: PointOffsetType,
+}
+
+pub struct AppendableMmapMultiDenseVectorStorage<T: PrimitiveVectorElement + 'static> {
+    vectors: ChunkedMmapVectors<T>,
+    offsets: ChunkedMmapVectors<MultivectorMmapOffset>,
+    deleted: DynamicMmapFlags,
+    distance: Distance,
+    multi_vector_config: MultiVectorConfig,
+    deleted_count: usize,
+}
+
+pub fn open_appendable_memmap_multi_vector_storage(
+    path: &Path,
+    dim: usize,
+    distance: Distance,
+    multi_vector_config: MultiVectorConfig,
+    stopped: &AtomicBool,
+) -> OperationResult<Arc<AtomicRefCell<VectorStorageEnum>>> {
+    let storage = open_appendable_memmap_multi_vector_storage_impl::<VectorElementType>(
+        path,
+        dim,
+        distance,
+        multi_vector_config,
+        stopped,
+    )?;
+
+    Ok(Arc::new(AtomicRefCell::new(
+        VectorStorageEnum::MultiDenseAppendableMemmap(Box::new(storage)),
+    )))
+}
+
+pub fn open_appendable_memmap_multi_vector_storage_impl<T: PrimitiveVectorElement>(
+    path: &Path,
+    dim: usize,
+    distance: Distance,
+    multi_vector_config: MultiVectorConfig,
+    stopped: &AtomicBool,
+) -> OperationResult<AppendableMmapMultiDenseVectorStorage<T>> {
+    create_dir_all(path)?;
+
+    let vectors_path = path.join(VECTORS_DIR_PATH);
+    let offsets_path = path.join(OFFSETS_DIR_PATH);
+    let deleted_path = path.join(DELETED_DIR_PATH);
+
+    let vectors = ChunkedMmapVectors::open(&vectors_path, dim)?;
+    let offsets = ChunkedMmapVectors::open(&offsets_path, 1)?;
+
+    let num_vectors = vectors.len();
+
+    let deleted: DynamicMmapFlags = DynamicMmapFlags::open(&deleted_path)?;
+
+    let mut deleted_count = 0;
+
+    for i in 0..num_vectors {
+        if deleted.get(i) {
+            deleted_count += 1;
+        }
+        check_process_stopped(stopped)?;
+    }
```

Comment:
I'm not sure if we have the same logic in a different place, but we might want to optimize this with a more efficient call for bit counting.

Reply:
It's a similar approach like in dense appendable mmap storage:
https://github.com/qdrant/qdrant/blob/dev/lib/segment/src/vector_storage/dense/appendable_mmap_dense_vector_storage.rs#L81
I guess better fix it later (in two places) if we notice that storage loading is a bottleneck

Reply:
> if we notice that storage loading is a bottleneck

Well, we are trying to profile slow loading storage now. I did notice this when using mmap's, though I'm not sure if it's related to this specifically.

Reply:
I have explored this idea here: https://github.com/qdrant/qdrant/pull/4176

If we merge it before this PR, we can simply change this to `deleted.count_flags()`.

Reply:
The above has been merged, so this one can be updated.

Reply:
Thanks for your great improvement. Changed here and use your counting now